### PR TITLE
separate DensityMatrixMP into its own measurement process

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Today, `qml.state()` returns a `StateMP` with no wires set, while `qml.density_matrix()` returns a `StateMP` with the wires of the subsystem being traced out. The new device implementation would benefit greatly from being able to assign wires to a state measurement, and also it just makes sense for these to be separate things.

**Description of the Change:**
Adds a new `DensityMatrixMP` that inherits from `StateMP` (not the parent `StateMeasurement` so that `isinstance(mp, StateMP)` still returns True for density matrix measurements).

I literally copy-pasted the `StateMP` class, then deleted any density-matrix-related things from the `StateMP` class, and deleted and state-related things from the `DensityMatrixMP` class.

**Benefits:**
- Density matrix is its own measurement process!
- We can assign wires to a `StateMP` without it thinking it represents a density matrix.

**Possible Drawbacks:**
This can also be viewed as a benefit (see above, no code changes outside module needed), but `isinstance(my_density_mp, StateMP)` returns true. If you want to know that your measurement _is_ a state measurement, you can do `type(mp) is StateMP`.

**UPDATE for reviewers**:
Responding to a comment, I've changed `StateMP` to not accept `wires` in its init, so now `DensityMatrixMP` directly calls `StateMeasurement.__init__(self, wires=wires, id=id)` to skip its parent (StateMP) which would now fail if given wires. I call it directly because this is the most pythonic way to call a grandparent's method (the other option being `super(StateMP, self).__init__` but I think `StateMeasurement` is a more reliable and responsible choice than "whoever StateMP's parent is"). Plus, I'll be removing that change when StateMP supports wires.

### CODE REVIEW

File: pennylane/measurements/state.py [hunk 1]
```diff
@@ -122,17 +122,16 @@ def circuit():
         with a compatible device.
     """
     wires = Wires(wires)
-    return StateMP(wires=wires)
+    return DensityMatrixMP(wires=wires)
 
 
 class StateMP(StateMeasurement):
```
Comment (albi3ro) [inline pennylane/measurements/state.py:128]: Should we forbid `wires` on the non-density matrix `StateMP`  for the moment? Then add a dedicated story in for `StateMP` can accept wires.

File: pennylane/measurements/state.py [hunk 2]
```diff
@@ -122,23 +122,21 @@ def circuit():
...
     Args:
-        wires (.Wires): The wires the measurement process applies to.
-            This can only be specified if an observable was not provided.
         id (str): custom label given to a measurement instance, can be useful for some applications
             where the instance has to be identified
     """
 
-    def __init__(self, wires: Optional[Wires] = None, id: Optional[str] = None):
-        super().__init__(wires=wires, id=id)
+    def __init__(self, *, id: Optional[str] = None):
```
Comment (rmoyard) [inline pennylane/measurements/state.py:138]: Why is the star necessary here?

**Review by rmoyard — APPROVED** [review]
Thanks, it is a nice one! Just one comment

File: pennylane/measurements/state.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/measurements/state.py:138]: not _super_ necessary, but without it, a user can do `StateMP([0])` and they might think it's the wires, but it'll actually set the `id`. This way, they get the error right away. Anyway, it'll be removed before the release when I add wires to StateMP :p

**Pull request merged**